platform-sdk/swirlds-platform-core/src/main/java/com/swirlds/platform/system/events/EventMetadata.java

```diff
@@ -0,0 +1,275 @@
+import com.swirlds.common.crypto.Hash;
+import com.swirlds.common.platform.NodeId;
+import com.swirlds.platform.system.SoftwareVersion;
+import com.swirlds.platform.system.transaction.TransactionWrapper;
+import edu.umd.cs.findbugs.annotations.NonNull;
+import edu.umd.cs.findbugs.annotations.Nullable;
+import java.time.Instant;
+import java.util.List;
+import java.util.Objects;
+import java.util.stream.Stream;
+
+/**
+ * Metadata for an event that can be derived from the event itself
+ */
+public class EventMetadata extends AbstractHashable {
+    /**
+     * the software version of the node that created this event.
+     */
+    private final SoftwareVersion softwareVersion;
+
+    /**
+     * ID of this event's creator (translate before sending)
+     */
+    private final NodeId creatorId;
+
+    /**
+     * the self parent event descriptor
+     */
+    private final EventDescriptorWrapper selfParent;
+
+    /**
+     * the other parents' event descriptors
+     */
+    private final List<EventDescriptorWrapper> otherParents;
+
+    /** a combined list of all parents, selfParent + otherParents */
+    private final List<EventDescriptorWrapper> allParents;
+
+    private final long generation;
+
+    /**
+     * creation time, as claimed by its creator
+     */
+    private final Instant timeCreated;
+
+    /**
+     * list of transactions
+     */
+    private final List<TransactionWrapper> transactions;
+    /**
+     * The event descriptor for this event. Is not itself hashed.
+     */
+    private EventDescriptorWrapper descriptor;
+
+    /**
+     * Create a EventMetadata object
+     *
+     * @param softwareVersion the software version of the node that created this event.
+     * @param creatorId       ID of this event's creator
+     * @param selfParent      self parent event descriptor
+     * @param otherParents    other parent event descriptors
+     * @param timeCreated     creation time, as claimed by its creator
+     * @param transactions    list of transactions included in this event instance
+     */
+    public EventMetadata(
+            @NonNull final SoftwareVersion softwareVersion,
+            @NonNull final NodeId creatorId,
+            @Nullable final EventDescriptorWrapper selfParent,
+            @NonNull final List<EventDescriptorWrapper> otherParents,
+            @NonNull final Instant timeCreated,
+            @NonNull final List<EventTransaction> transactions) {
+
+        Objects.requireNonNull(transactions, "The transactions must not be null");
+        this.softwareVersion = Objects.requireNonNull(softwareVersion, "The softwareVersion must not be null");
+        this.creatorId = Objects.requireNonNull(creatorId, "The creatorId must not be null");
+        this.selfParent = selfParent;
+        Objects.requireNonNull(otherParents, "The otherParents must not be null");
+        otherParents.forEach(Objects::requireNonNull);
+        this.otherParents = otherParents;
```

Comment:
Stylistic Nit, the same can be achieved by using [List#copyOf](https://docs.oracle.com/javase/10/docs/api/java/util/List.html#copyOf(java.util.Collectionl))
```suggestion
        this.otherParents = List.copyOf(otherParents);
```
It will throw NullPointerException if `otherParents` or any of its elements are null.

Reply:
changed